src/main/flight/pid.c

```diff
@@ -445,19 +445,47 @@ STATIC_UNIT_TESTED FAST_CODE_NOINLINE float pidLevel(int axis, const pidProfile_
 #ifdef USE_GPS_RESCUE
     angle += gpsRescueAngle[axis] / 100; // ANGLE IS IN CENTIDEGREES
 #endif
-    angle = constrainf(angle, -levelAngleLimit, levelAngleLimit);
     const float errorAngle = angle - ((attitude.raw[axis] - angleTrim->raw[axis]) / 10.0f);
+    float angleFF = 0;
+    //angle loop feedforward
+#ifdef USE_FEEDFORWARD
+    if (newRcFrame){
+        const float rcCommandDelta = fabsf(getRcCommandDelta(axis));
+        float angleFfInput = pidRuntime.angleFeedforward[axis];
+        uint8_t duplicateCount = getFeedforwardDuplicateCount(axis);
+        if (rcCommandDelta) {
+            const float rxRateHz = 1e6f / getCurrentRxRefreshRate();
+            angleFfInput = (angle - pidRuntime.previousAngle[axis]) * rxRateHz;
+            if (duplicateCount) {
+                    angleFfInput /= duplicateCount + 1;
+                }
+        } else {
+            if (duplicateCount > 1) {
+                angleFfInput = 0;
+            }
+        }
+        //angle feedforward filter at RCrate
+        pidRuntime.angleFeedforward[axis] = pt2FilterApply(&pidRuntime.angleFeedforwardPt2[axis], angleFfInput);
+        pidRuntime.previousAngle[axis] = angle;
+    }
+    //angle feedforward filter at the PID loop rate
+    angleFF = pt3FilterApply(&pidRuntime.angleFeedforwardPt3[axis], pidRuntime.angleFeedforward[axis]);
+#endif
     if (FLIGHT_MODE(ANGLE_MODE) || FLIGHT_MODE(GPS_RESCUE_MODE)) {
         // ANGLE mode - control is angle based
-        const float setpointCorrection = errorAngle * pidRuntime.levelGain;
-        currentPidSetpoint = pt3FilterApply(&pidRuntime.attitudeFilter[axis], setpointCorrection);
+        rawSetpoint = errorAngle * pidRuntime.levelGain + angleFF * pidRuntime.levelFFGain;
     } else {
         // HORIZON mode - mix of ANGLE and ACRO modes
         // mix in errorAngle to currentPidSetpoint to add a little auto-level feel
-        const float setpointCorrection = errorAngle * pidRuntime.horizonGain * horizonLevelStrength;
-        currentPidSetpoint += pt3FilterApply(&pidRuntime.attitudeFilter[axis], setpointCorrection);
+        rawSetpoint += (errorAngle * horizonLevelStrength + angleFF * pidRuntime.levelFFGain) * pidRuntime.horizonGain;
     }
-    return currentPidSetpoint;
+    //logging
+    if (axis == FD_ROLL) {
+        DEBUG_SET(DEBUG_ANGLE_CONTROLLER, 0, lrintf(angle * 10.0f));
+        DEBUG_SET(DEBUG_ANGLE_CONTROLLER, 1, lrintf(errorAngle * 10.0f));
+        DEBUG_SET(DEBUG_ANGLE_CONTROLLER, 2, lrintf(angleFF * 10000.0f));
```

Comment:
Maybe it will be easiest to interpret the debug if it carries the direct contribution to final setpoint.  

For angle mode, perhaps:
- instead of logging errorAngle, log errorAngle * pidRuntime.levelGain, and 
- instead of logging angleFF, log angleFF * pidRuntime.levelFFGain.  

Unfortunately Horizon Mode uses different gains.

However maybe it could be refactored so that we can log the final two elements that make up rawSetpoint.